File: quartermaster-code-runner/.claude/worktrees/nice-shockley/src/quartermaster_code_runner/runtime/node/mcp-client.js
/**
 * MCP Stdio Client - Executes MCP server commands and calls tools via stdio transport
 *
 * Environment variables:
 * - MCP_COMMAND: The command to start the MCP server (e.g., "npx @modelcontextprotocol/server-slack")
 * - MCP_OPERATION: The operation to perform: "list_tools" or "call_tool" (default: "call_tool")
 * - MCP_TOOL_NAME: The name of the tool to call (required for call_tool operation)
 * - MCP_TOOL_ARGUMENTS: JSON string of arguments to pass to the tool
 * - MCP_ENV_*: Additional environment variables to pass to the MCP server (prefix stripped)
 */

const { spawn } = require("child_process");
const readline = require("readline");

const MCP_PROTOCOL_VERSION = "2024-11-05";
const INITIALIZE_TIMEOUT = 60000; // 60 seconds for slow package resolution
const TOOL_CALL_TIMEOUT = 120000; // 2 minutes

class McpStdioClient {
  constructor(command, serverEnv = {}) {
    this.command = command;
    this.serverEnv = serverEnv;
    this.process = null;
    this.messageId = 0;
    this.pendingRequests = new Map();
    this.initialized = false;
    this.serverStderrLines = [];
    this.processExited = false;
    this.exitCode = null;
  }

  async start() {
    return new Promise((resolve, reject) => {
      const env = {
        ...process.env,
        ...this.serverEnv,
      };

      delete env.MCP_COMMAND;
      delete env.MCP_TOOL_NAME;
      delete env.MCP_TOOL_ARGUMENTS;

      this.process = spawn(this.command, {
        stdio: ["pipe", "pipe", "pipe"],
        env,
        shell: true,
      });

      this.process.on("error", (err) => {
        reject(new Error(`Failed to start MCP server: ${err.message}`));
      });

      this.process.on("exit", (code, signal) => {
        this.processExited = true;
        this.exitCode = code;

        const stderrTail = this.serverStderrLines.slice(-20).join("\n");
        const msg = `MCP server exited with code ${code}${signal ? `, signal ${signal}` : ""}.\nServer output:\n${stderrTail}`;

        for (const [id, { reject: rej }] of this.pendingRequests) {
          rej(new Error(msg));
        }
        this.pendingRequests.clear();

        if (!this.initialized) {
          reject(new Error(msg));
        }
      });

      const rl = readline.createInterface({
        input: this.process.stdout,
        crlfDelay: Infinity,
      });

      rl.on("line", (line) => {
        try {
          const message = JSON.parse(line);
          this.handleMessage(message);
        } catch (e) {
          console.error("[MCP Server]", line);
        }
      });

      this.process.stderr.on("data", (data) => {
        const text = data.toString().trimEnd();
        this.serverStderrLines.push(text);
        console.error("[MCP Server]", text);
      });

      setTimeout(() => {
        if (this.process && !this.process.killed && !this.processExited) {
          resolve();
        }
      }, 500);
    });
  }

  handleMessage(message) {
    if (message.id !== undefined && this.pendingRequests.has(message.id)) {
      const { resolve, reject } = this.pendingRequests.get(message.id);
      this.pendingRequests.delete(message.id);

      if (message.error) {
        reject(
          new Error(message.error.message || JSON.stringify(message.error)),
        );
      } else {
        resolve(message.result);
      }
    }
  }

  sendRequest(method, params = {}) {
    return new Promise((resolve, reject) => {
      const id = ++this.messageId;
      const request = {
        jsonrpc: "2.0",
        id,
        method,
        params,
      };

      this.pendingRequests.set(id, { resolve, reject });
      this.process.stdin.write(JSON.stringify(request) + "\n");
    });
  }

  sendNotification(method, params = {}) {
    const notification = {
      jsonrpc: "2.0",
      method,
      params,
    };
    this.process.stdin.write(JSON.stringify(notification) + "\n");
  }

  async initialize() {
    const result = await Promise.race([
      this.sendRequest("initialize", {
        protocolVersion: MCP_PROTOCOL_VERSION,
        capabilities: {},
        clientInfo: {
          name: "quartermaster-code-runner",
          version: "1.0.0",
        },
      }),
      new Promise((_, reject) =>
        setTimeout(
          () => reject(new Error("Initialize timeout")),
          INITIALIZE_TIMEOUT,
        ),
      ),
    ]);

    // Send initialized notification
    this.sendNotification("notifications/initialized");
    this.initialized = true;

    return result;
  }

  async listTools() {
    const result = await this.sendRequest("tools/list", {});
    return result.tools || [];
  }

  async callTool(name, arguments_) {
    const result = await Promise.race([
      this.sendRequest("tools/call", {
        name,
        arguments: arguments_,
      }),
      new Promise((_, reject) =>
        setTimeout(
          () => reject(new Error("Tool call timeout")),
          TOOL_CALL_TIMEOUT,
        ),
      ),
    ]);

    // Parse content from result
    const content = result.content || [];
    const textParts = [];

    for (const item of content) {
      if (item.type === "text") {
        textParts.push(item.text || "");
      } else if (item.type === "image") {
        textParts.push(`[Image: ${item.mimeType || "unknown"}]`);
      } else if (item.type === "resource") {
        textParts.push(`[Resource: ${item.uri || "unknown"}]`);
      }
    }

    return textParts.join("\n");
  }

  async close() {
    if (this.process && !this.process.killed) {
      this.process.kill();
    }
  }
}

async function main() {
  const command = process.env.MCP_COMMAND;
  const operation = process.env.MCP_OPERATION || "call_tool";
  const toolName = process.env.MCP_TOOL_NAME;
  const toolArgumentsJson = process.env.MCP_TOOL_ARGUMENTS || "{}";

  if (!command) {
    console.error("Error: MCP_COMMAND environment variable is required");
    process.exit(1);
  }

  if (operation === "call_tool" && !toolName) {
    console.error(
      "Error: MCP_TOOL_NAME environment variable is required for call_tool operation",
    );
    process.exit(1);
  }

  let toolArguments;
  try {
    toolArguments = JSON.parse(toolArgumentsJson);
  } catch (e) {
    console.error("Error: MCP_TOOL_ARGUMENTS must be valid JSON");
    process.exit(1);
  }

  // Extract MCP_ENV_* variables for the server
  const serverEnv = {};
  for (const [key, value] of Object.entries(process.env)) {
    if (key.startsWith("MCP_ENV_")) {
      serverEnv[key.substring(8)] = value;
    }
  }

  const client = new McpStdioClient(command, serverEnv);

  try {
    console.error(`Starting MCP server: ${command}`);
    await client.start();

    console.error("Initializing MCP connection...");
    await client.initialize();

    if (operation === "list_tools") {
      console.error("Listing tools...");
      const tools = await client.listTools();
      console.log(JSON.stringify(tools));
    } else {
      console.error(`Calling tool: ${toolName}`);
      const result = await client.callTool(toolName, toolArguments);
      console.log(result);
    }

    await client.close();
    process.exit(0);
  } catch (error) {
    console.error("MCP Error:", error.message);
    if (client.serverStderrLines.length > 0) {
      console.error("--- MCP Server stderr ---");
      for (const line of client.serverStderrLines) {
        console.error(line);
      }
      console.error("--- End server stderr ---");
    }
    await client.close();
    process.exit(1);
  }
}

main();

File: quartermaster-code-runner/.claude/worktrees/nice-shockley/src/quartermaster_code_runner/runtime/python/mcp-client.py
#!/usr/bin/env python3
"""
MCP Stdio Client - Executes MCP server commands and calls tools via stdio transport

Environment variables:
- MCP_COMMAND: The command to start the MCP server (e.g., "uvx mcp-server-fetch")
- MCP_OPERATION: The operation to perform: "list_tools" or "call_tool" (default: "call_tool")
- MCP_TOOL_NAME: The name of the tool to call (required for call_tool operation)
- MCP_TOOL_ARGUMENTS: JSON string of arguments to pass to the tool
- MCP_ENV_*: Additional environment variables to pass to the MCP server (prefix stripped)
"""

import json
import os
import subprocess
import sys
import threading
from typing import Any

MCP_PROTOCOL_VERSION = "2024-11-05"
INITIALIZE_TIMEOUT = 60  # 60 seconds for slow package resolution
TOOL_CALL_TIMEOUT = 120


class McpStdioClient:
    def __init__(self, command: str, server_env: dict[str, str] | None = None):
        self.command = command
        self.server_env = server_env or {}
        self.process: subprocess.Popen | None = None
        self.message_id = 0
        self.pending_requests: dict[int, dict] = {}
        self.initialized = False
        self._lock = threading.Lock()
        self._response_event = threading.Event()
        self._reader_thread: threading.Thread | None = None
        self._stderr_thread: threading.Thread | None = None
        self._exit_thread: threading.Thread | None = None
        self._last_response: dict | None = None
        self._server_stderr_lines: list[str] = []
        self._process_exited = False
        self._exit_code: int | None = None

    def start(self):
        env = os.environ.copy()
        env.update(self.server_env)

        for key in ["MCP_COMMAND", "MCP_TOOL_NAME", "MCP_TOOL_ARGUMENTS"]:
            env.pop(key, None)

        self.process = subprocess.Popen(
            self.command,
            shell=True,
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            env=env,
            text=True,
            bufsize=1,
        )

        self._reader_thread = threading.Thread(target=self._read_responses, daemon=True)
        self._reader_thread.start()
        self._stderr_thread = threading.Thread(target=self._read_stderr, daemon=True)
        self._stderr_thread.start()
        self._exit_thread = threading.Thread(target=self._monitor_exit, daemon=True)
        self._exit_thread.start()

    def _monitor_exit(self):
        if not self.process:
            return
        self._exit_code = self.process.wait()
        self._process_exited = True
        self._response_event.set()

    def _read_stderr(self):
        if not self.process or not self.process.stderr:
            return
        for line in self.process.stderr:
            line = line.rstrip("\n")
            self._server_stderr_lines.append(line)
            print(f"[MCP Server] {line}", file=sys.stderr)

    def _read_responses(self):
        if not self.process or not self.process.stdout:
            return

        for line in self.process.stdout:
            line = line.strip()
            if not line:
                continue
            try:
                message = json.loads(line)
                self._handle_message(message)
            except json.JSONDecodeError:
                print(f"[MCP Server stdout] {line}", file=sys.stderr)

    def _handle_message(self, message: dict):
        msg_id = message.get("id")
        if msg_id is not None and msg_id in self.pending_requests:
            with self._lock:
                self._last_response = message
                self._response_event.set()

    def _send_request(
        self, method: str, params: dict | None = None, timeout: int = 30
    ) -> Any:
        self.message_id += 1
        request_id = self.message_id

        request = {
            "jsonrpc": "2.0",
            "id": request_id,
            "method": method,
            "params": params or {},
        }

        with self._lock:
            self.pending_requests[request_id] = {}
            self._response_event.clear()
            self._last_response = None

        if self._process_exited:
            raise RuntimeError(
                f"MCP server exited with code {self._exit_code} before {method} could be sent"
            )

        if self.process and self.process.stdin:
            self.process.stdin.write(json.dumps(request) + "\n")
            self.process.stdin.flush()

        if not self._response_event.wait(timeout):
            raise TimeoutError(f"{method} timed out after {timeout}s")

        if self._process_exited and self._last_response is None:
            stderr_tail = "\n".join(self._server_stderr_lines[-20:])
            raise RuntimeError(
                f"MCP server process exited with code {self._exit_code} during {method}.\n"
                f"Server output:\n{stderr_tail}"
            )

        with self._lock:
            response = self._last_response
            self.pending_requests.pop(request_id, None)

        if response and "error" in response:
            error = response["error"]
            raise Exception(error.get("message", str(error)))

        return response.get("result") if response else None

    def _send_notification(self, method: str, params: dict | None = None):
        notification = {
            "jsonrpc": "2.0",
            "method": method,
            "params": params or {},
        }

        if self.process and self.process.stdin:
            self.process.stdin.write(json.dumps(notification) + "\n")
            self.process.stdin.flush()

    def initialize(self) -> dict:
        result = self._send_request(
            "initialize",
            {
                "protocolVersion": MCP_PROTOCOL_VERSION,
                "capabilities": {},
                "clientInfo": {
                    "name": "quartermaster-code-runner",
                    "version": "1.0.0",
                },
            },
            timeout=INITIALIZE_TIMEOUT,
        )

        self._send_notification("notifications/initialized")
        self.initialized = True
        return result

    def list_tools(self) -> list[dict]:
        result = self._send_request("tools/list", {})
        return result.get("tools", []) if result else []

    def call_tool(self, name: str, arguments: dict | None = None) -> str:
        result = self._send_request(
            "tools/call",
            {"name": name, "arguments": arguments or {}},
            timeout=TOOL_CALL_TIMEOUT,
        )

        content = result.get("content", []) if result else []
        text_parts = []

        for item in content:
            if item.get("type") == "text":
                text_parts.append(item.get("text", ""))
            elif item.get("type") == "image":
                text_parts.append(f"[Image: {item.get('mimeType', 'unknown')}]")
            elif item.get("type") == "resource":
                text_parts.append(f"[Resource: {item.get('uri', 'unknown')}]")

        return "\n".join(text_parts)

    def close(self):
        if self.process:
            self.process.terminate()
            try:
                self.process.wait(timeout=5)
            except subprocess.TimeoutExpired:
                self.process.kill()


def main():
    command = os.environ.get("MCP_COMMAND")
    operation = os.environ.get("MCP_OPERATION", "call_tool")
    tool_name = os.environ.get("MCP_TOOL_NAME")
    tool_arguments_json = os.environ.get("MCP_TOOL_ARGUMENTS", "{}")

    if not command:
        print(
            "Error: MCP_COMMAND environment variable is required",
            file=sys.stderr,
        )
        sys.exit(1)

    if operation == "call_tool" and not tool_name:
        print(
            "Error: MCP_TOOL_NAME environment variable is required for call_tool operation",
            file=sys.stderr,
        )
        sys.exit(1)

    try:
        tool_arguments = json.loads(tool_arguments_json)
    except json.JSONDecodeError:
        print("Error: MCP_TOOL_ARGUMENTS must be valid JSON", file=sys.stderr)
        sys.exit(1)

    server_env = {}
    for key, value in os.environ.items():
        if key.startswith("MCP_ENV_"):
            server_env[key[8:]] = value

    client = McpStdioClient(command, server_env)

    try:
        print(f"Starting MCP server: {command}", file=sys.stderr)
        client.start()

        print("Initializing MCP connection...", file=sys.stderr)
        client.initialize()

        if operation == "list_tools":
            print("Listing tools...", file=sys.stderr)
            tools = client.list_tools()
            print(json.dumps(tools))
        else:
            print(f"Calling tool: {tool_name}", file=sys.stderr)
            result = client.call_tool(tool_name or "", tool_arguments)
            print(result)

        client.close()
        sys.exit(0)
    except Exception as error:
        print(f"MCP Error: {error}", file=sys.stderr)
        if client._server_stderr_lines:
            print("--- MCP Server stderr ---", file=sys.stderr)
            for line in client._server_stderr_lines:
                print(line, file=sys.stderr)
            print("--- End server stderr ---", file=sys.stderr)
        client.close()
        sys.exit(1)


if __name__ == "__main__":
    main()
